File: Lab-3/read_mapping_2.py
from calendar import c
from easyinput import read_line

'''
    Pseudocode:

    suffixes = dict()
    for i in |sequence|: suffixes[i] = sequence[i:]
    {print(k[0]) for k in sorted(suffixes.items(), key=lambda item: item[1])}


'''

def bubble_sort(unsorted,reverse):

    n = len(unsorted)
    if reverse: 
        for i in range(n-1):
            for j in range(n-1,i, -1):
                if unsorted[j] < unsorted[j-1]: 
                    #Don't need to create a tmp variable to exchange 2 values.
                    unsorted[j], unsorted[j-1] = unsorted[j-1],unsorted[j]
    else:
        for i in range(n-1):
            for j in range(n-1,i, -1):
                if unsorted[j] > unsorted[j-1]: 
                    #Don't need to create a tmp variable to exchange 2 values.
                    unsorted[j], unsorted[j-1] = unsorted[j-1],unsorted[j]
    return unsorted


def suffix_array(sequence):

    #suffixes = dict()
    suff = list()
    #suff_dup = list()
    
    for i in range(0, len(sequence)): suff.append(sequence[i:])
        #suff_dup.append(sequence[i:])
    
    suff_dup = suff.copy()

    suff = bubble_sort(suff, True)

    for i in range(0, len(sequence)): print(suff_dup.index(suff[i])+1)


suffix_array(read_line())

File: Lab-3/read_mapping_2.cpp
#include <string>
#include <iostream>
#include <vector>
#include <map>

using namespace std;

/*
    Pseudocode:

    FUNCTION SUFFIX-ARRAY ():
        suffixes = dict()
        for i in |sequence|: suffixes[i] = sequence[i:]
        {print(k[0]) for k in sorted(suffixes.items(), key=lambda item: item[1])}
*/

void bubble_sort ( vector < pair < int , string > >& T ) {
    int n = T.size();
    for ( int i=0; i < n-1; ++i) //From i=0 to length of the vector -1
        for ( int j = n-1; j>i ; --j ) //From j = size -1 to j bigger than i 
            if ( T[j].second < T[j-1].second) swap ( T[j] , T[j-1]); //If position 2 smaller than position 1, swap them.
}

void printVector(vector<pair<int,string>>& v){
    for (int i = 0; i < v.size(); ++i) cout << v.at(i).first << endl;
}

void suffix_array(string seq) {

    vector< pair< int, string > > suff;

    //Storing data into the map.
    for ( int i=0 ; i < seq.length() ; ++i ) suff.push_back(make_pair( i+1, seq.substr(i) ));
    //Sorting the data.
    bubble_sort(suff);

    printVector(suff);


}


int main(){

    string seq;
    getline(cin, seq);
    suffix_array(seq);
    return 0;

}